src/solver/algorithm/strong/puzzle.rs

```diff
@@ -9,95 +9,233 @@ use anyhow::{Context, Result};
         .context("Failed to initialize volatile database.")?;
 
-    bfs(&mut db, game)
+    reverse_bfs_solver(&mut db, game)
         .context("Failed solving algorithm execution.")?;
 
     Ok(())
 }
 
-
-fn bfs<G, D>(game: &G, db: &mut D)
+fn reverse_bfs_solver<G, D>(db: &mut D, game: &G) -> Result<()>
 where
-    G: DTransition<State> + Bounded<State> + SimpleSum<N>,
-    D: KVStore<RecordBuffer>,
+    G: DTransition<State>
+        + Bounded<State>
+        + ClassicPuzzle
+        + Extensive<1>
+        + Game,
+    D: KVStore,
 {
-    let end_states = discover_end_states_helper(db, game);
+    // Get end states and create frontiers
+    let end_states = discover_child_counts(db, game)?;
+
+    let mut winning_queue: VecDeque<State> = VecDeque::new();
+    let mut losing_queue: VecDeque<State> = VecDeque::new();
+    for end_state in end_states {
+        match ClassicPuzzle::utility(game, end_state) {
+            SimpleUtility::WIN => winning_queue.push_back(end_state),
+            SimpleUtility::LOSE => losing_queue.push_back(end_state),
+            SimpleUtility::TIE => Err(SolverViolation {
+                name: "PuzzleSolver".to_string(),
+                hint: format!(
+                    "Primitive end position cannot have utility TIE
+                              for a puzzle"
+                ),
+            })?,
+            SimpleUtility::DRAW => Err(SolverViolation {
+                name: "PuzzleSolver".to_string(),
+                hint: format!(
+                    "Primitive end position cannot have utility DRAW
+                              for a puzzle"
+                ),
+            })?,
+        }
+        // Add ending state utility and remoteness to database
+        update_db_record(db, end_state, game.utility(end_state), 0, 0)?;
+    }
 
-    for state in end_states {
-        let mut buf = RecordBuffer::new()
-            .context("Failed to create placeholder record.")?;
-        buf.set_remoteness(0)
-            .context("Failed to set remoteness for end state.")?;
-        db.put(state, &buf);
+    // Perform BFS on winning states
+    while let Some(state) = winning_queue.pop_front() {
+        let buf = RecordBuffer::from(db.get(state).unwrap())?;
+        let child_remoteness =
+            RecordBuffer::from(db.get(state).unwrap())?.get_remoteness();
 
-        bfs_state(db, game, state); 
+        for parent in game.retrograde(state) {
+            let child_count =
+                RecordBuffer::from(db.get(parent).unwrap())?.get_child_count();
+            if child_count > 0 {
+                winning_queue.push_back(parent);
+                update_db_record(
+                    db,
+                    parent,
+                    SimpleUtility::WIN,
+                    1 + child_remoteness,
+                    0,
+                )?;
+            }
+        }
     }
+
+    // Perform BFS on losing states, where remoteness is the longest path to a
```

Comment:
Please avoid using comments -- feel free to refactor this into a new function (and add this comment as a docstring if necessary).